### PR TITLE
Copernicus Essential Climate Variables for assessing climate variability

### PR DESCRIPTION
**Pull request by annefou**
- added python + xml and 2 tests
- requires CDS API key to work locally.

### CODE REVIEW

**Comment by annefou** [comment]
@bgruening 

Hi Björn,

**NOT URGENT!!!**

Whenever you have time, I would really appreciate to get your inputs/review for this new tool *essential_climate_variables*. (Learning progress, I made a new branch instead of updating the master branch...).

Thanks.

File: tools/essential_climate_variables/ecv_retrieve.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            tar.close()
+
+
+if __name__ == '__main__':
+    warnings.filterwarnings("ignore")
+    parser = argparse.ArgumentParser()
+
+    remove_apikey = False
+    current_pwd = os.getcwd()
+    if 'GALAXY_APIRC_KEY' in os.environ and not os.path.isfile('.cdsapirc'):
```
Comment (bgruening) [inline tools/essential_climate_variables/ecv_retrieve.py]: Where should this `.cdsapirc` be? Currently, it needs to be in next to the script, isn't it?

Comment (bgruening) [reply tools/essential_climate_variables/ecv_retrieve.py]: I would make the "GALAXY_APIRC_KEY" longer and more meaningful.

File: tools/essential_climate_variables/ecv_retrieve.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        apikey = open(".cdsapirc", "w+")
```
Comment (bgruening) [inline tools/essential_climate_variables/ecv_retrieve.py]: Use `with  open(".cdsapirc", "w+") as apikey` here ...

File: tools/essential_climate_variables/ecv_retrieve.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        apikey.write("url: https://cds.climate.copernicus.eu/api/v2\n")
+        apikey.write("key: " + os.environ['GALAXY_APIRC_KEY'])
+        apikey.close()
+        remove_apikey = True
+        os.environ['HOME'] = current_pwd
```
Comment (bgruening) [inline tools/essential_climate_variables/ecv_retrieve.py]: Set this inside the Galaxy XML wrapper, like `HOME='pwd'`

File: tools/essential_climate_variables/ecv_retrieve.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                'month': self.month,
+                'area': self.area,
+                'format': self.format,
+                'product_type': self.product_type,
+                'time_aggregation': self.time_aggregation,
+                          },
+                self.outputfile)
+
+    def checktar(self):
+        ofile = open(self.outputfile, 'rb')
```
Comment (bgruening) [inline tools/essential_climate_variables/ecv_retrieve.py]: use python ``with open` statement.

File: tools/essential_climate_variables/ecv_retrieve.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        is_grib = ofile.read(4)
+        ofile.close()
+        if (is_grib == b'GRIB' and self.format == 'tgz'):
+            # we create a tgz to be consistent
+            newfilename = tempfile.NamedTemporaryFile()
+            print(newfilename.name)
```
Comment (bgruening) [inline tools/essential_climate_variables/ecv_retrieve.py]: debug output?

File: tools/essential_climate_variables/ecv_retrieve.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    args = parser.parse_args()
+
+    p = ECV(args.archive, args.variable, args.product_type,
+            args.year, args.month, args.time_aggregation, args.area,
+            args.format, args.output, args.verbose)
+    p.retrieve()
+    p.checktar()
+    # remove api key file if it was created
+    if remove_apikey and os.getcwd() == current_pwd:
+        os.remove(current_pwd + '/.cdsapirc')
```
Comment (bgruening) [inline tools/essential_climate_variables/ecv_retrieve.py]: use, os.path.join() here.

File: tools/essential_climate_variables/README.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+
+# Copernicus Essential Climate Variables for assessing climate variability
+
+This tool allows you to retrieve Copernicus **Essential Climate Variables** for assessing climate variability from the [Copernicus Climate Data Store](https://cds.climate.copernicus.eu/#!/home).
+
+Copernicus ECV for assessing climate variability is available from the Copernicus portal at https://cds.climate.copernicus.eu/cdsapp#!/dataset/ecv-for-climate-change?tab=overview
+
+To be able to retrieve data, you would need to register and get your [CDS API key](https://cds.climate.copernicus.eu/api-how-to).
```
Comment (bgruening) [inline tools/essential_climate_variables/README.md]: Please add an explanation to about the ENV var and the hidden API key.

File: tools/essential_climate_variables/essential_climate_variables.xml [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+             <option value="2010">2010</option>
+             <option value="2011">2011</option>
+             <option value="2012">2012</option>
+             <option value="2013">2013</option>
+             <option value="2014">2014</option>
+             <option value="2015">2015</option>
+             <option value="2016">2016</option>
+             <option value="2017">2017</option>
+             <option value="2018">2018</option>
+             <option value="2019" selected="true">2019</option>
```
[the remark this answers is not in the document]
Comment (bgruening) [reply tools/essential_climate_variables/essential_climate_variables.xml]: Another option is to generate this list out of a flat file. You can put next to the XML a config file and the select will read from it. This way a admin could add enter new years without changing the XML.

Obviously the best way is the date-picker :)

**Comment by bgruening** [comment]
> (Learning progress, I made a new branch instead of updating the master branch...)

Awesome! :)

I'm a little bit puzzels how the tests can turn green if you don't have a API key here.

**Comment by annefou** [comment]
So fast! Thanks for all your inputs! That's so helpful to get a review.

The tests are running OK because I added my own API key as an environment variable in travis. If you do not have any API keys, all the tests fail.

**Comment by bgruening** [comment]
> The tests are running OK because I added my own API key as an environment variable in travis. If you do not have any API keys, all the tests fail.

Very nice! :)

File: tools/essential_climate_variables/ecv_retrieve.py [hunk 9]
```diff
@@ -66,13 +65,13 @@ def retrieve(self):
                 self.outputfile)
 
     def checktar(self):
-        ofile = open(self.outputfile, 'rb')
-        is_grib = ofile.read(4)
-        ofile.close()
+        is_grib = False
+        with open(self.outputfile, 'rb') as ofile:
+            is_grib = ofile.read(4)
+            ofile.close()
```
Comment (bgruening) [inline tools/essential_climate_variables/ecv_retrieve.py]: the cool thing with the with statement is that you do not need to clean the file handle :)

https://preshing.com/20110920/the-python-with-statement-by-example/

File: tools/essential_climate_variables/essential_climate_variables.xml [hunk 10]
```diff
@@ -6,6 +6,7 @@
         <requirement type="package" version="1.9.6">cdo</requirement>
     </requirements>
     <command detect_errors="exit_code"><![CDATA[
+##       HOME=`pwd`  &&
```
Comment (bgruening) [inline tools/essential_climate_variables/essential_climate_variables.xml]: ?

File: tools/essential_climate_variables/essential_climate_variables.xml [hunk 11]
```diff
@@ -153,14 +153,6 @@ the corresponding anomalies.
...
 
-**What it does**
-----------------
-
-   This tool retrieves Essential Climate Variables (ECVs) from Copernicus
-   Climate Data Store.
-
-**Usage**
-
 ::
```
Comment (bgruening) [inline tools/essential_climate_variables/essential_climate_variables.xml]: Do you think the rest is useful?

File: tools/essential_climate_variables/essential_climate_variables.xml [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+  See climatedata.wmo.int/ for more information on Essential Climate Variables.
+
+.. class:: infomark
+
+   It allows to retrieve estimates of Essential Climate Variables (ECVs)
+   derived from Earth observations.
+
+    ]]></help>
+    <citations>
```
Comment (bgruening) [inline tools/essential_climate_variables/essential_climate_variables.xml]: Should Copernicus be cited maybe?

Comment (annefou) [reply tools/essential_climate_variables/essential_climate_variables.xml]: Yes. I just did not find any publications on it; except webpages or powerpoint presentation so I was not sure how to cite Copernicus!

Comment (bgruening) [reply tools/essential_climate_variables/essential_climate_variables.xml]: Ok, that's fine then I guess.

Comment (bgruening) [reply tools/essential_climate_variables/essential_climate_variables.xml]: What about this here: https://emergency.copernicus.eu/mapping/ems/cite-copernicus-ems-mapping-portal

Comment (annefou) [reply tools/essential_climate_variables/essential_climate_variables.xml]: It is related to emergency management service and not climate service and I did not find anything clear at http://climate.copernicus.eu

Comment (bgruening) [reply tools/essential_climate_variables/essential_climate_variables.xml]: Ok, then let's ignore this for the moment :)

**Pull request merged**